### PR TITLE
Refactor preprocessing files.

### PR DESCRIPTION
**Pull request by MarkAZhang**
# Description

- Removed the `reshape_mosquito_per_image` file, which had many common methods with `preprocess_mosquito_image` and added the `allocate_one_mosquito` functionality to `preprocess_mosquito_image`. When the user provides the optional argument `individual_mosquito_dir`, all individual mosquito images will be placed in that directory.
- Moved all preprocessing methods into a separate file `preprocessing.py`.
- Added an additional command `debug_preprocessing_single_image`, which runs the preprocessing on a single image and outputs all images to an `output_dir`. This function also outputs the intermediate `binary_by_saturation_filter` image, which will help us debug any changes we make there.
- Added a `main.py` function that allows us to call the functions from Python, instead of via command line. (I got tired of trying to type these super long commands in the terminal)
- Generalized the `allocate_one_mosquito` image a little bit (renamed it to `get_subimage`)
- Also fixed a couple of issues that were preventing me from running the code, including changing the version of Pandas and changing the function signature on `_, contours, _ = cv2.findContours`. My version of `findContours` only outputs 2 parameters.

### CODE REVIEW

**Comment by pranathivemuri** [comment]
I like having command lines so people can access and run the code themselves without having to import functions. Its just easier to give them a command and ask them to copy paste instead of giving them a program. I had to just rename the folder names and copy paste the commands for different folders. So, I didn’t type it multiple times

For opencv and pandas - i am not going to change the versions. Please create a conda environment woth environment.yml file requirements

I am going to merge the rest of the parts of your changes that are good refactor requests and merge them into my branch

Thanks!

**Comment by MarkAZhang** [comment]
- It's still possible to use the command line directly. I just added an additional way to call it from code. I can rename main.py to test.py so it's clear that it's not the intended way to run the library (and we can remove that file before we publish the library). I find the code interface a little easier to use personally. Let me know if that works.

- I'm a little confused by the `environment.yml` versions comment. I feel like `requirements.txt` should have versions that match what's in the code. I changed the versions because when I tried to run `pip install -r requirements.txt` and then run the Python code I got some errors. Perhaps you can explain a little more.

File: debug_preprocessing_single_image.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    # Save a clean copy that doesn't have bounding boxes drawn on it.
+    color_mosquito_image_copy = color_mosquito_image.copy()
+
+    file_name, file_extension = os.path.basename(args.input_file).split(".")
+
+    binary_mosquito_image = binarize_by_saturation_filter(
+        color_mosquito_image, args.min_saturation, args.max_saturation)
+
+    cv2.imwrite(
```
Comment (pranathivemuri) [inline debug_preprocessing_single_image.py]: binary image should be saved as x, y shaped image with no rgb channels

File: example.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+from preprocess_mosquito_images import preprocess_mosquito_images, get_preprocess_mosquito_images_argparser
+from debug_preprocessing_single_image import debug_preprocessing_single_image, get_debug_preprocessing_single_image_argparser
+
+PREPROCESS_MOSQUITO_IMAGES_ARGPARSER = get_preprocess_mosquito_images_argparser()
+DEBUG_PREPROCESSING_SINGLE_IMAGE_ARGPARSER = get_debug_preprocessing_single_image_argparser()
+
+# Example of calling the library directly from Python. Feel free to modify according to your directory structure.
+# You can also call the library directly from the command line.
+def process_all_mosquiti_images():
```
Comment (pranathivemuri) [inline example.py]: mosquito spelling

File: requirements.txt [hunk 3]
```diff
@@ -8,7 +8,6 @@ click<7,>=6.7
 natsort==6.2.0
 opencv-python==4.1.0.25
 scipy==1.2.1
-natsort==6.0.0
 matplotlib==3.0.3
 itkwidgets==0.14.0
-pandas<0.21
+pandas==0.25.2
```
Comment (pranathivemuri) [inline requirements.txt]: I changed these lines now in my branch

Comment (pranathivemuri) [reply requirements.txt]: also changed opencv version.

File: example.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+from preprocess_mosquito_images import preprocess_mosquito_images, get_preprocess_mosquito_images_argparser
+from debug_preprocessing_single_image import debug_preprocessing_single_image, get_debug_preprocessing_single_image_argparser
+
+PREPROCESS_MOSQUITO_IMAGES_ARGPARSER = get_preprocess_mosquito_images_argparser()
+DEBUG_PREPROCESSING_SINGLE_IMAGE_ARGPARSER = get_debug_preprocessing_single_image_argparser()
+
+# Example of calling the library directly from Python. Feel free to modify according to your directory structure.
```
Comment (pranathivemuri) [inline example.py]: sorry I missed this! thanks for the comments and example

File: preprocessing.py [hunk 5]
```diff
@@ -208,102 +151,50 @@ def draw_bb_labels_around_contours(
...
-        required=False, type=int, default=MAX_MOSQUITO_AREA)
-    parser.add_argument(
-        "--input_image_format",
-        help="input image data format", required=True, type=str)
-
-    args = parser.parse_args()
+def get_subimage(
+        color_mosquito_image, bounding_box):
+    """
+    Returns a list of single mosquito images of TILE_SIZE
```
Comment (pranathivemuri) [inline preprocessing.py]: change docstring

File: preprocessing.py [hunk 6]
```diff
@@ -208,102 +151,50 @@ def draw_bb_labels_around_contours(
...
-    output_dir = args.output_dir
-    if os.path.exists(output_dir) is False:
-        os.mkdir(output_dir)
 
-    # For each image, fing all bounding boxes and write them to the csv file
-    for path in images:
-        image = cv2.imread(path)
-        color_mosquito_image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+def extract_subimages_by_bounding_boxes(image, bounding_boxes):
+    """
```
Comment (pranathivemuri) [inline preprocessing.py]: correct the docstring to be like other functions descriptions with input args and return args.

**Comment by pranathivemuri** [comment]
@MarkAZhang thanks for the refactor, I have a few comments regarding docstrings that didn't look right to me. fix them and we will merge this onto the other PR.

**Pull request merged**